### DISCUSSION TITLE
Help with modeling a fiber reinforced soft actuator using the rigidify function

### DISCUSSION THREAD

**Discussion by tolstoys19** [SofaDefrost/SoftRobots#192] [Help - Building / Configuring] [answered]
Dear All,

Right now i am trying to model a soft actuator that has a fiber reinforcement wrapped around it in order to restrict radial expansion.
i saw the example of rigidification used on the tripod tutorial. I am currently struggling with defining the group  indices and the frames required for the function.  In here you can access my code and meshes:
my code is based on the PneunetGripper tutorial and the tripod tutorial

Google Drive:

https://drive.google.com/drive/folders/1MU1hpfG_mX78J8h6voJho1vyKY76oI4N?usp=sharing

There was a similar discussion thread in here

https://www.sofa-framework.org/community/forum/topic/fiber-reinforced-soft-robot/

As I am entering the coordinates of the boxes manually I am inputting the indices directly into the groupindices variable. 
For the frames variable I have read the definition attached on the Rigidify function documentation but I am still not quite clear on its meaning. 

My current understanding is that the frames specify the coordinates of the points on the outer surface where the fibers would be located on the real actuator(the points I want to rigidify). I don't understand however how could I retrieve this information.
According to one of the answers in the previous forum the data of the frames corresponds quaternion of all  the frames however I am not sure what would this for my model. 
With my current approach I am getting the following error:
```
Python exception:
  ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

At:
  C:/Users/owner/Desktop/SOFA_robosoft2022_python-3.8_Windows/plugins/STLIB/lib/python3/site-packages\stlib3\physics\mixedmaterial\rigidification.py(74): mfilter
  C:/Users/owner/Desktop/SOFA_robosoft2022_python-3.8_Windows/plugins/STLIB/lib/python3/site-packages\stlib3\physics\mixedmaterial\rigidification.py(81): Rigidify
  rigidificationtry3.py(21): __rigidify
  rigidificationtry3.py(36): createScene
```
I would be deeply grateful if you could help me with this issues.
I am very new to SOFA so I apologize in advance if my question is very basic
Best Regards,
Pablo

**Comment by EulalieCoevoet** [comment, marked as answer]
Hi @tolstoys19,

Sorry for the very late answer. 
You want to rigidify groups of points, i.e. you want each group of points to form a rigid frame. A rigid frame in SOFA is composed of seven values, the first three are the position of the frame, and the other four are a quaternion which gives the orientation of the frame. 

You have three options here: 
- by default if you don't specify any values for the variable `frames`, the position of each frame will be set as the geometric center of the group of points you gave, and the orientation will be `[0, 0, 0, 1]`.
- if the geometric center is a good option for you but you want to specify the orientation, then you can set `frames = [[qx, qy, qz, qw], ...]`.
- finally if you want to set both the position and orientation of each frame `frames=[[x, y, z, qx, qy, qz, qw], ...]`

**Reply by tolstoys19** [reply]
@EulalieCoevoet 

Dear Eulalie. Thank you for your help, I was able to solve this problem some months ago but I noticed that I needed a very fine meshing to get this method to work and even then sometimes I wouldn't get all the results that I wanted. Right now I am trying another approach using stiff springs to model the fibers but I have still not succeeded, I have uploaded the question in this thread:

https://github.com/sofa-framework/sofa/discussions/3587

I would be really happy if you could give it a check.

Best Regards,
Pablo Tortós
